### PR TITLE
Check Example File Exists

### PR DESCRIPTION
**Pull request by ambidexterich**
I had to do some massaging to get the example file to resolve to the correct directory but you might have an easier way to do it.

### CODE REVIEW

File: loaders/utils/getProps.js [hunk 1]
```diff
@@ -22,16 +24,29 @@ const getDoctrineTags = documentation => {
...
 };
 
+const validateExampleFile = (baseFile, exampleFile) => {
+	const componentFile = baseFile || '';
+	if (typeof exampleFile !== 'string') {
+		return false;
+	}
+
+	// Remove newline characters included from react-docgen
+	const cleanFilePath = exampleFile.replace(/\n/g, '');
```
Comment (sapegin) [inline loaders/utils/getProps.js]: `.trim()` should be enough here. Also I’d do it outside the function so you could remove the same code when you print a warning message.

File: loaders/utils/getProps.js [hunk 2]
```diff
@@ -49,14 +64,21 @@ module.exports = function getProps(doc) {
...
 		doc.tags = getDoctrineTags(documentation);
 
 		doc.description = highlightCode(removeDoclets(doc.description));
-
-		if (doc.doclets.example) {
+		if (exampleFileExists) {
 			doc.example = requireIt(`!!${examplesLoader}!${doc.doclets.example}`);
 			delete doc.doclets.example;
+		} else if (doc.doclets.example) {
+			/* eslint-disable no-console */
```
Comment (sapegin) [inline loaders/utils/getProps.js]: eslint-disable-nextline

File: loaders/utils/getProps.js [hunk 3]
```diff
@@ -49,14 +64,21 @@ module.exports = function getProps(doc) {
...
+			console.warn(
+				`${chalk.bold(doc.doclets.example.toString().replace(/\n/g, ''))} is an invalid filepath.`
```
Comment (sapegin) [inline loaders/utils/getProps.js]: I’d not use styling here, it may make #472 harder.

File: loaders/utils/getProps.js [hunk 4]
```diff
@@ -22,16 +23,35 @@ const getDoctrineTags = documentation => {
 	return _.groupBy(documentation.tags, 'title');
 };
 
+const validateExampleFile = (baseFile, exampleFile) => {
+	// In case there is no basefile such as running tests but exampleFile might still exist
```
Comment (sapegin) [inline loaders/utils/getProps.js]: Could we pass a name from tests? I’d remove both conditions.

File: loaders/utils/getProps.js [hunk 5]
```diff
@@ -22,16 +23,35 @@ const getDoctrineTags = documentation => {
...
+	if (typeof exampleFile !== 'string') {
+		return false;
+	}
+
+	const cleanFilePath = exampleFile.trim();
+	const exampleFilepath = path.resolve(path.dirname(componentFile), cleanFilePath);
+	const doesFileExist = fs.existsSync(exampleFilepath);
+
+	if (!doesFileExist) {
+		// eslint-disable-next-line
```
Comment (sapegin) [inline loaders/utils/getProps.js]: This way you’re disabling ESLint entirely for the next line — there may be other warnings.

File: loaders/utils/getProps.js [hunk 6]
```diff
@@ -22,16 +23,35 @@ const getDoctrineTags = documentation => {
...
+		console.warn(`${cleanFilePath} is an invalid filepath.`);
```
Comment (sapegin) [inline loaders/utils/getProps.js]: If file doesn’t exist it doesn’t mean the path is invalid. To make it useful for the user I’d try something like this:

> An example file FILEPATH defined in COMPONENTPATH component not found.

File: loaders/utils/getProps.js [hunk 7]
```diff
@@ -22,16 +23,35 @@ const getDoctrineTags = documentation => {
 	return _.groupBy(documentation.tags, 'title');
 };
 
+const validateExampleFile = (baseFile, exampleFile) => {
```
Comment (sapegin) [inline loaders/utils/getProps.js]: I’d call it something like `isExampleFileExists` and do only that. Right now it does too many things:

1. Check if example defined.
2. Check if file exists.
3. Prints a warning.
4. Do other magic like trimming file names and checking if component path was passed or not.

**Comment by ambidexterich** [comment]
I've implemented all of your feedback. The warning is a stopgap until https://github.com/styleguidist/react-styleguidist/issues/497. I can't start using styleguidist in our product until this one is fixed.

**Comment by sapegin** [comment]
The initial version was closer to what I’d merge unfortunately ;-|

File: loaders/utils/getProps.js (hunk 5 above)
Comment (ambidexterich) [inline loaders/utils/getProps.js]: Good catch

File: loaders/utils/getProps.js (hunk 7 above)
Comment (ambidexterich) [inline loaders/utils/getProps.js]: The reason for the check that example file is a string and ensure that the base file exists is that `path.resolve` fails if either one is not a string like when @example is true and passed from a test. Would you prefer that moved to the body of `getProps`?

Comment (sapegin) [inline loaders/utils/getProps.js]: Definitely! ;-)

File: loaders/utils/getProps.js (hunk 4 above)
Comment (ambidexterich) [inline loaders/utils/getProps.js]: Updated

**Comment by ambidexterich** [comment]
Submitted another attempt.

**Comment by sapegin** [comment]
Cool, thanks!

**Pull request merged**